x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java

```diff
@@ -151,13 +168,129 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+    static void updateExecutionInfoWithFieldCapsResults(
+        EsqlExecutionInfo executionInfo,
+        Map<String, StringBuilder> clusterAndResolvedIndices,
+        List<FieldCapabilitiesFailure> failures
+    ) {
+        Set<String> clustersWithoutFieldCapsResponses = new HashSet<>(executionInfo.getClusterAliases());
+        for (Map.Entry<String, StringBuilder> entry : clusterAndResolvedIndices.entrySet()) {
+            final String clusterAlias = entry.getKey();
+            final EsqlExecutionInfo.Cluster cluster = executionInfo.getCluster(clusterAlias);
+            assert cluster != null : "All cluster aliases should have already been inserted into ExecutionInfo. Missing: " + clusterAlias;
+            executionInfo.swapCluster(
+                clusterAlias,
+                // overwrite the existing unresolved index expression with the resolved indices
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(cluster).setIndexExpression(entry.getValue().toString()).build()
+            );
+            clustersWithoutFieldCapsResponses.remove(clusterAlias);
+        }
+
+        /*
+         * These are clusters in the original request that are not present in the field-caps response. They were
+         * specified with an index or indices that do not exist, so the search on that cluster is done.
+         * Mark it as complete with 0 shards searched and 0 took time.
+         */
+        for (String c : clustersWithoutFieldCapsResponses) {
+            executionInfo.swapCluster(
+                c,
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
+                    .setTook(new TimeValue(0))
+                    .setTotalShards(0)
+                    .setSuccessfulShards(0)
+                    .setSkippedShards(0)
+                    .setFailedShards(0)
+                    .build()
+            );
+        }
+
+        if (failures != null) {
+            Set<String> unavailableClusters = new HashSet<>();
+            for (FieldCapabilitiesFailure failure : failures) {
+                if (isRemoteUnavailableException(failure.getException())) {
+                    for (String indexExpression : failure.getIndices()) {
+                        if (indexExpression.indexOf(RemoteClusterAware.REMOTE_CLUSTER_INDEX_SEPARATOR) > 0) {
+                            unavailableClusters.add(parseClusterAlias(indexExpression));
+                        }
+                    }
+                }
+            }
+            for (String clusterAlias : unavailableClusters) {
+                executionInfo.swapCluster(
+                    clusterAlias,
+                    (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SKIPPED).build()
+                );
+                // TODO: follow-on PR will set SKIPPED status when skip_unavailable=true and throw an exception when skip_unavailable=false
+            }
```

Comment:
I was not able to find a good way to write a test against this scenario. I did try to set up a test in CrossClustersQueryIT where I created a third remote cluster and then took all of its data nodes offline, resulting in SKIPPED status, but the test was very flaky, so I removed it. I did test this scenario manually several times.

Reply:
Could you try testing with MockTransportService#addConnectBehavior?